### PR TITLE
8330275: Crash in XMark::follow_array

### PR DESCRIPTION
**Pull request by ashu-mehra**
This PR addresses the issue in ZGC where the number of address offset bits can go beyond the limit imposed by the encoding scheme in mark stack, thereby causing the encoding to fail.
Encoding of partial array offset in mark stack requires that the max address bit be no more than 46 bit. ~~But the current mechanism to probe maximum address offset bits on aarch64, riscv and ppc platforms can return value larger that 44 bits. This patch sets the maximum address offset bits to 44.~~

~~I have updated the generational mode to avoid subtracting 3 bits from the maximum address offset bit probed by the system, as the generational mode does not use multi-mapping.~~

~~I have also updated the code to set MarkPartialArrayMinSizeShift dynamically depending on the number of address offset bits used. This would avoid running into such problem again if in future maximum address offset bits is increased beyond 44.~~

~~For some reason (that I can't comprehend from the code) the existing implementation for probing the max addressable bit for ppc in non-generation ZGC is very different from other platforms and from generational mode as well. I have kept the existing implementation as is and just fixed it to ensure it does not return value greater than 44 bits.~~

Testing: ~~test/hotspot/jtreg/gc/z and test/hotspot/jtreg/gc/x on x86~~ tier1, tier2 and tier3 on aarch64 using fastdebug build with options JTREG="EXTRA_PROBLEM_LISTS=ProblemList-zgc.txt;JAVA_OPTIONS=-XX:+UseZGC -XX:+ZVerifyOops;JOBS=4"  (as per the suggestion in [JDK-8330275](https://bugs.openjdk.org/browse/JDK-8330275?focusedId=14667864&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14667864))

Update: Striked out the changes that are not relevant now that it is only doing a point fix for aarch64

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330275](https://bugs.openjdk.org/browse/JDK-8330275): Crash in XMark::follow_array (**Bug** - P2)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [1529968d](https://git.openjdk.org/jdk/pull/18941/files/1529968da6a2a22d536776a60c3c642acd1f3a2e)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [1529968d](https://git.openjdk.org/jdk/pull/18941/files/1529968da6a2a22d536776a60c3c642acd1f3a2e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18941/head:pull/18941` \
`$ git checkout pull/18941`

Update a local copy of the PR: \
`$ git checkout pull/18941` \
`$ git pull https://git.openjdk.org/jdk.git pull/18941/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18941`

View PR using the GUI difftool: \
`$ git pr show -t 18941`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18941.diff">https://git.openjdk.org/jdk/pull/18941.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18941#issuecomment-2075812559)

### CODE REVIEW

**Comment by ashu-mehra** [comment]
I am currently trying to get access to aarch64 system and run the tests test/hotspot/jtreg/gc/z and test/hotspot/jtreg/gc/x.
I would appreciate if some one can also test the ppc and riscv changes as I don't have access to such systems.

**Review by stefank — CHANGES_REQUESTED** [review]
Hi @ashu-mehra,

Thanks for fixing this issue.

There's a number of changes style changes I would like to make to make sure that the code looks more inline with what the rest of the ZGC code looks like. But before we start with that I would like to request that we skip making the changes to marking stack code and limit the changes to only the probing code. Doing so will make it easier to get this fix reviewed and delivered.

**Comment by ashu-mehra** [comment]
@stefank I am trying to understand the reason behind your suggestion to remove the changes in marking stack code. Are they not correct or is it that they don't belong to this PR?
Anyway I have removed them from this PR.

**Comment by tstuefe** [comment]
> > @stefank I am trying to understand the reason behind your suggestion to remove the changes in marking stack code. Are they not correct or is it that they don't belong to this PR? Anyway I have removed them from this PR.
> 
> To me, it was not bleeding obvious that they were the right thing to do, and given other changes that doesn't follow the grown ZGC coding style, I wanted suggest a way forward for you to get this bug fixed, with less resistance from us ZGC developers/maintainers. That was the reasoning.

I agree with Stefan. I would keep the patch as minimal as possible to make it easier to follow the actual error that has been fixed, and to make it easier for backporters to decide what to downport. 

Code cleanups can happen in a separate RFE.

Ashu, are the other platforms actually broken? If yes, which ones? If a platform is not broken, I would defer touching it up to a separate cleanup RFE. Again because of patch clarity.

**Comment by ashu-mehra** [comment]
I agree from the point of view of backporting, point-fix is all we need in this PR.

@tstuefe  As for the other platforms (riscv and ppc), looking at their code they seem to be broken in the same way as aarch64 but then the problem only happens if the user runs with > 1TB heap size with more than 48 addressable bits.
Again, in the spirit of "do not touch if it is not broken", I am fine if we restrict the change to just aarch64.

@tstuefe @stefank please let me know if you agree with just doing the point-fix to aarch64.

**Comment by tstuefe** [comment]
> I agree from the point of view of backporting, point-fix is all we need in this PR.
> 
> @tstuefe As for the other platforms (riscv and ppc), looking at their code they seem to be broken in the same way as aarch64 but then the problem only happens if the user runs with > 1TB heap size with more than 48 addressable bits. Again, in the spirit of "do not touch if it is not broken", I am fine if we restrict the change to just aarch64.
> 
> @tstuefe @stefank please let me know if you agree with just doing the point-fix to aarch64.

Absolutely. We can do any platform testing on other platforms and cleanups in subsequent RFEs.

**Comment by ashu-mehra** [comment]
Sorry for the long absence on this PR. I have updated the PR to just do a point fix for aarch64. I have also done tier1, tier2 and tier3 tests on aarch64.

**Review by stefank — APPROVED** [review]
Looks good.

FWIW, the 128TB and 64GB numbers are just confusing when we are talking about a bit position. If the 46th bit succeeds the usable address range is 128TB, and the 46th bit will account for 64TB out of those 128TB. I wouldn't mind at all if we just ripped out the mentions of 64TB and  64GB here.

**Comment by ashu-mehra** [comment]
> I wouldn't mind at all if we just ripped out the mentions of 64TB and 64GB here.

Done

**Comment by ashu-mehra** [comment]
@tstuefe does this look ok?

File: src/hotspot/cpu/aarch64/gc/x/xGlobals_aarch64.cpp [hunk 1]
```diff
@@ -142,9 +142,9 @@
 //  * 63-48 Fixed (16-bits, always zero)
 //
 
-// Default value if probing is not implemented for a certain platform: 128TB
-static const size_t DEFAULT_MAX_ADDRESS_BIT = 47;
-// Minimum value returned, if probing fails: 64GB
```
Comment (tstuefe) [inline src/hotspot/cpu/aarch64/gc/x/xGlobals_aarch64.cpp:147]: any reason you removed the comment for MINIMUM_MAX_ADDRESS_BIT?

**Review by tstuefe — APPROVED** [review]
Looks good, question inline.

**Comment by ashu-mehra** [comment]
As the last commit is a trivial change to add the comments back, I am not requesting new review and integrating it as is.

File: src/hotspot/cpu/aarch64/gc/z/zAddress_aarch64.cpp [hunk 2]
```diff
@@ -36,9 +36,11 @@
 #include <sys/mman.h>
 #endif // LINUX
 
-// Default value if probing is not implemented for a certain platform: 128TB
-static const size_t DEFAULT_MAX_ADDRESS_BIT = 47;
-// Minimum value returned, if probing fails: 64GB
+// Default value if probing is not implemented for a certain platform
+// Max address bit is restricted by implicit assumptions in the code, for instance
+// the bit layout of XForwardingEntry or Partial array entry (see XMarkStackEntry) in mark stack
```
Comment (stefank) [inline src/hotspot/cpu/aarch64/gc/z/zAddress_aarch64.cpp:41]: This comment was copy-n-pasted without updating the names to ZForwardingEntry and ZMarkStackEntry.

**Pull request closed without merging**